kythe/go/extractors/config/preprocessor/modifier/pom_xml_modifier.go

```diff
@@ -50,10 +56,128 @@ func PreProcessPomXML(pomXMLFile string) error {
+func hasCompilerPlugin(doc *etree.Document) (bool, error) {
+	for _, p := range doc.FindElements("//project/build/plugins/plugin") {
+		a := p.FindElement("artifactId")
+		if a.Text() == "maven-compiler-plugin" {
+			// Now, we either check for the existence of correctly-specified
+			// elements for groupId, version, and
+			// configuration/[source|target], or we check for the complete
+			// absence of all those elements.  For the former we check
+			// validity, and for the latter we just assume that there's some
+			// well-formed parent pluginManagement specifying the values.
+			if len(p.ChildElements()) == 1 {
+				// Assume that there's some well-formed parent
+				// pluginManagement specifying the values.
+				// TODO(danielmoy): handle the case where pom.xml plugin is
+				// incorrectly specified, by looking at all of the pom.xml
+				// files in the whole project wholistically.  This might be
+				// out of scope - we might insist they be fixed upstream.
+				return true, nil
+			}
+			return true, wellformedCompilerPlugin(p)
+		}
+	}
+	for _, p := range doc.FindElements("//project/build/pluginManagement/plugins/plugin") {
+		a := p.FindElement("artifactId")
+		if a.Text() == "maven-compiler-plugin" {
+			// For pluginManagement, we do need a well-specified set of
+			// sibling elements.
+			return true, wellformedCompilerPlugin(p)
+		}
+	}
+
+	// If we didn't find anything, we can proceed with modification.
+	return false, nil
+}
+
+// Check to see if the plugin fits the minimum supported versions we've
+// tested:
+// <plugin>
+//   <artifactId>maven-compiler-plugin</artifactId>
+//   <version>3.7.0</version>
+//   <configuration>
+//     <source>1.8</source>
+//     <target>1.8</target>
+//   </configuration>
+// </plugin>
+// TODO(danielmoy): see if we can relax some of these restrictions.
+func wellformedCompilerPlugin(el *etree.Element) error {
+	v := el.FindElement("version")
+	if v == nil {
+		return fmt.Errorf("missing maven-compiler-plugin version")
+	}
+	vt, err := semvar(v.Text())
+	if err != nil || vt.earlierThan(versiontuple{[]int{3, 7, 0}}) {
+		return fmt.Errorf("unsupported maven-compiler-plugin version: %s, need 3.7.0+", v.Text())
+	}
+
+	if err := checkJavaVersion(el, "source"); err != nil {
+		return err
+	}
+	if err := checkJavaVersion(el, "target"); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func checkJavaVersion(el *etree.Element, kind string) error {
+	k := el.FindElement(fmt.Sprintf("configuration/%s", kind))
+	if k == nil {
+		return fmt.Errorf("mising %s", kind)
+	}
+	if v, err := javaVersion(k.Text()); err != nil || v.earlierThan(versiontuple{[]int{1, 7}}) {
+		// TODO(#3075): Should test an upper bound here too.
+		return fmt.Errorf("unsupported %s version: %s, need 1.7+", kind, k.Text())
+	}
+	return nil
+}
+
+// Some container that supports semvar (3.7.0) and java versions (1.6) for
+// comparison purposes only.
+type versiontuple struct {
```

Comment:
This can probably just be `type versionTuple []int`.

Reply:
ah yea.  Originally I had this only for semvar and it made sense with the struct, but after refactor to handle both semvar and java version in the same thing, the struct makes very little sense.

Thanks.